feat: Add new Vue FilePicker from `@nextcloud/dialogs` and use it by default

susnux opened the pull request:
* Part of #38926
  * Resolves: #39186
  * Resolves: #39187
  * Resolves: #39188

## Summary
*(Still providing the legacy one until the Vue FilePicker is out of beta.)*

Basically just use the new component from the dialogs package, see information in the PR https://github.com/nextcloud-libraries/nextcloud-dialogs/pull/878

### Screenshots

![image](https://github.com/nextcloud-libraries/nextcloud-dialogs/assets/1855448/1f14373c-735a-4ed2-a476-4bdd93f19deb)

https://github.com/nextcloud-libraries/nextcloud-dialogs/assets/1855448/3f5ec8fe-ec11-4308-824f-5dca90c72752

And on smaller screens:
![image](https://github.com/nextcloud-libraries/nextcloud-dialogs/assets/1855448/59e2c06f-8c48-45a9-8277-818a7b19d77c)

## TODO

- [ ] See list on the dialogs PR: https://github.com/nextcloud-libraries/nextcloud-dialogs/pull/878
- Especially https://github.com/nextcloud/server/issues/39185 will come after the first beta.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

susnux commented:
/backport to stable27

skjnldsv commented on core/src/OC/dialogs.js:
```diff
@@ -246,7 +250,7 @@ const Dialogs = {
 	 * @param {Function} [options.filter] filter function for advanced filtering
 	 */
-	filepicker: function(title, callback, multiselect, mimetypeFilter, modal, type, path, options) {
+	legacyFilepicker: function(title, callback, multiselect, mimetypeFilter, modal, type, path, options) {
```
deprecate message ? :)

susnux replied:
sure :)

AndyScherzinger commented:
> (Still providing the legacy one until the Vue FilePicker is out of beta.)

but files will use the new one right-away?

susnux commented:
> but files will use the new one right-away?

Correct by default the new one is used everywhere.

szaimen reviewed (dismissed):
A bit feedback from design perspective:
1. Why is there an empty space between home button and plus button?
2. Why is there a checkbox to select folders?
![image](https://github.com/nextcloud/server/assets/42591237/54fb132d-d008-4c6d-bef2-31997feb41f4)
3. On mobile, the seearch and multiselect looks distorted. They should probably be on different lines?
![image](https://github.com/nextcloud/server/assets/42591237/83a977ed-ed79-4167-b86b-f3973fbf013d)

AndyScherzinger commented:
/backport to stable27

jancborchardt reviewed (dismissed):
Nice! :) Only 2 additional things besides @szaimen’s points:
- [ ] The loading skeleton screen elements are too dark on the background. We had a similar issue in Mail or elsewhere, I thought it was fixed in the components?
- [ ] The "Modified" column should be left-aligned like it is now in files. Only "Size" is right-aligned as it has numbers which can extend to the left, but modified is numbers + text.

susnux commented:
Thank you for your feedback :)

@szaimen 
> Why is there an empty space between home button and plus button?
I think this is an issue with the nextcloud vue library as this is from the NcBreadcrumbs component.

> Why is there a checkbox to select folders?
Currently there is always checkboxes shown (if multiselect is enabled then there is also a select all one in the table header).
Which brings up the question if this should be changed.

---

@jancborchardt 
>  The loading skeleton screen elements are too dark on the background.

You are right, on darkmode I did not noticed it that much but I agree we should make them lighter.

> The "Modified" column should be left-aligned like it is now in files.
Thank you, will change this :)

szaimen commented:
Thanks @susnux! :)

> > Why is there an empty space between home button and plus button?
> 
> I think this is an issue with the nextcloud vue library as this is from the NcBreadcrumbs component.

I see, I can create an issue for this if you want :)

> > Why is there a checkbox to select folders?
> 
> Currently there is always checkboxes shown (if multiselect is enabled then there is also a select all one in the table header).
> Which brings up the question if this should be changed.

Yes, I think so :)

---

Please dont forget my third point :)

> On mobile, the seearch and multiselect looks distorted. They should probably be on different lines?

susnux commented:
> Please dont forget my third point :)
Yes :sweat_smile: 

I thought I had already fixed this but seems like the styling was not updated.

susnux committed:
feat: Add new Vue FilePicker from @nextcloud/dialogs and use it by default.

Still providing the legacy one until the Vue FilePicker is out of beta.
Pin beta releases so we do not get version conflicts.

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(cypress): Fix the user theming cypress test for the new filepicker

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

nimishavijay commented:
Super cool! :) Some small points from my side as well:
- [ ] The heading for the entire dialog is somewhat off center, would it be possible to center it? :) Similar to Talk/Mail/Files settings dialogs 
![image](https://github.com/nextcloud/server/assets/52440189/20119457-727f-490b-bff3-6c5a063d20dc)

- [ ] The icons for each file/folder also seem a bit big, we could change it to a 20px [MDI folder icon](https://fonts.google.com/icons?selected=Material%20Icons%20Round%3Afolder%3A). This is different from the way it currently looks in Files, but we should change it in Files also after F2V.
- [ ] Small detail, in the header "Name" can be left aligned with the text in the rows below (currently it is left aligned with the icon)



> Currently there is always checkboxes shown (if multiselect is enabled then there is also a select all one in the table header).
Which brings up the question if this should be changed.

If only one file/folder can be selected (eg. while moving/copying a file) then it doesn't make much sense to show the checkboxes as they wouldn't do anything. However if multiple files/folders can be selected (eg. when adding attachments to an email) then the checkboxes could be shown to indicate selection, what do you think?

skjnldsv commented:
Maybe we can summarize all design requests in an issue in https://github.com/nextcloud-libraries/nextcloud-dialogs ?

susnux commented:
Comments of @jancborchardt are resolved. The third one of @szaimen also (the first one is probably a nextcloud-vue bug, the second one will be in a future beta release).

---
@nimishavijay Thank you very much :)

> Super cool! :) Some small points from my side as well:
> 
>     * [ ]  The heading for the entire dialog is somewhat off center, would it be possible to center it? :) Similar to Talk/Mail/Files settings dialogs

Well to be honest I implemented it according to the mockup in https://github.com/nextcloud/server/issues/38926#issuecomment-1632560834
We should (at a different location) discuss if we should adjust the settings dialog instead by using the new general purpose dialog component as the base.

>     * [ ]  The icons for each file/folder also seem a bit big, we could change it to a 20px [MDI folder icon](https://fonts.google.com/icons?selected=Material%20Icons%20Round%3Afolder%3A). This is different from the way it currently looks in Files, but we should change it in Files also after F2V.

Currently the same as the files app (even as the new f2v field app), so with Nextcloud 28 we can adjust this, what do you think?
Especially when previews instead of generic icons are shown.

But generally lets open a new issue on the dialogs library and do a design review session for the next iteration, would that work for you? :)

nimishavijay commented:
> Well to be honest I implemented it according to the mockup 

Ah! I see the confusion, in the mockup it is actually centered, but it just happens to align perfectly with the file list section :)
Would it be possible to just center it for the upcoming release, and address the other points for 28? :)

susnux committed:
chore: Remove unused FilePickerContext from behat tests

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

skjnldsv commented:
Rebasing...

skjnldsv commented:
Needs https://github.com/nextcloud/server/pull/39955

susnux committed:
fix(apps): Adjust apps for FilePicker API change

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
chore: Remove now unneeded filepicker styles

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(dialogs): Use FilePickerBuilder to create the file picker

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

skjnldsv commented:
If the cypress user test fails with `Request failed with status code 500`
It means the path given and returner by the picker is invalid :)

skjnldsv committed:
fix(theming): user background picker

Signed-off-by: John Molakvoæ <skjnldsv@protonmail.com>

skjnldsv commented:
Added a theming fix.
Now we need a new dialogs release and we should be good :rocket:

susnux committed:
chore: Build assets

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux commented:
should be ready to go?

skjnldsv commented:
Congrats!! :tada: :tada: :tada:
